### PR TITLE
Numark N4 remake

### PR DESCRIPTION
**Pull request by Swiftb0y**
I felt very limited by the old mapping so I created my fully fledged remake of the old mapping.
Features:
- seamless integration of crossfader orientation and PFL-buttons (work differently on the N4 than in Mixxx)
- all the necessary controls for looping
- support for up to 16 hotcues
- full control over the library navigation
- basic fx unit control
- switching Deck3&4 between PC- and line-in
- used components.js for better read- and maintainability
- all the usual/labeled controls on the N4 work
I tried to PR the mapping earlier this year but it still had too many bugs and I didn't have time to fix them back then, so I closed the PR.

### CODE REVIEW

**Comment by Be-ing** [comment]
Is the [wiki](https://mixxx.org/wiki/doku.php/numark_n4) up to date with your latest changes?

**Comment by Be-ing** [comment]
Ping @Swiftb0y, can you confirm that the wiki is up to date with the mapping before I review it?

**Comment by Swiftb0y** [comment]
Oh sorry. I forgot about this because of #1830 but I'll try to get around updating the wiki this evening.

**Comment by Swiftb0y** [comment]
@Be-ing turns out there wasn't much to change. The Wiki is up-to-date now.

**Comment by rryan** [comment]
Status? It'd be nice to have this in 2.2!

**Comment by Swiftb0y** [comment]
Its pretty much done but I recently discovered a "bug" where the hotcuepage doesn't reset when a new track gets loaded, So I want to update that before merging. I'll try to fix that today.

**Comment by Swiftb0y** [comment]
I don't think I will be able to solve that hotcuepage issue in time as it would probably require a change within the C++ source (which I'm not comfortable with this shortly before release, especially because I can't test for unintended side effects within other mappings). https://mixxx.zulipchat.com/#narrow/stream/113295-controller-mapping/topic/Binding.20to.20whenever.20a.20Deck.20loads.20a.20new.20track

**Comment by Swiftb0y** [comment]
I have narrowed down the problem and it seems that it is just a tiny tweak, Which won't affect any of the mixxx internals and only a small part of the script, so I'll probably will go that route of modifying the C++ source. I hope I can get around that before this weekend.

**Comment by Swiftb0y** [comment]
I already updated the wiki with my changes as well.

**Review by Be-ing — CHANGES_REQUESTED** [review]
Sorry for the delay reviewing this and thanks for your patience.

There are lots of places throughout the code that are a bit hard to read because the code is squeezed together. Please use spaces around operators and after commas. Also, please break `if` statements and functions across multiple lines.

File: res/controllers/Numark-N4-4channel-scripts.js [hunk 1]
```diff
@@ -0,0 +1,674 @@
...
+      }
+  }
+};
+
+NumarkN4.init = function (id) {
+  NumarkN4.rateRanges[0]=engine.getValue("[Channel1]","rateRange");
+  //PFL-Master-Mix & headGain are not being controlled by the controller,
+  //this sets them in engine to something reasonable so the user doesn't get confused if there is not sound.
+  engine.setParameter("[Master]","headGain",Math.max(engine.getParameter("[Master]","headGain"),0.5));
+  engine.setParameter("[Master]","headMix",0); //Headphone out = 100% pfl signal.
```
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/Numark-N4-4channel-scripts.js]: Ok, I guess I'll add a note about that in the wiki.

File: res/controllers/Numark-N4-4channel-scripts.js [hunk 2]
```diff
@@ -0,0 +1,674 @@
...
+
+  this.btnEffect1 = new components.Button({
+    midi: [0x90+channel,0x13,0xB0+channel,0x0B],
+    shift: function () {
+      this.group="[EffectRack1_EffectUnit1]";
+      this.type=components.Button.prototype.types.toggle;
+      this.inKey="group_[Channel"+channel+"]_enable";
+      this.outKey="group_[Channel"+channel+"]_enable";
+    },
+    unshift: function () {
```
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/Numark-N4-4channel-scripts.js]: I really don't want to change that for every single control. If you really prefer that, I'd add a section with coding guidelines specifically for component.js to the wiki

File: res/controllers/Numark-N4-4channel-scripts.js [hunk 3]
```diff
@@ -0,0 +1,674 @@
...
+      this.type=components.Button.prototype.types.push;
+      this.inKey="beatloop_activate";
+      this.outKey="beatloop_activate";
+    },
+  });
+  this.btnSample4 = new components.Button({
+    midi: [0x90+channel,0x16,0xB0+channel,0x0E],
+    shift: function () {
+      this.type=components.Button.prototype.types.toggle;
+      this.inKey="repeat";
```
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/Numark-N4-4channel-scripts.js]: I wasn't aware of reloop_andstop when I first wrote the mapping. I'll change that to reloop_andstop as it makes more sense and also matches the behavior of the loop button within the Skins.

File: res/controllers/Numark-N4-4channel-scripts.js [hunk 4]
```diff
@@ -0,0 +1,674 @@
...
+        theContainer.hotcueButtons[i].disconnect();
+        theContainer.hotcueButtons[i].number=number;
+        theContainer.hotcueButtons[i].outKey='hotcue_' + number + '_enabled';
+        theContainer.hotcueButtons[i].unshift(); // for setting inKey based on number property.
+        theContainer.hotcueButtons[i].connect();
+        theContainer.hotcueButtons[i].trigger();
+      }
+      if (displayFeedback) {
+        for (var i=0;i<4;++i) {
+          midi.sendShortMsg(0xB0+channel,0x0B+i,(i-this.hotCuePage)?0x00:0x7F);
```
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/Numark-N4-4channel-scripts.js]: The controller has a eight buttons arranged in two rows with four cells each. The second row is used to represent a set of four cuepoints. When you turn the encoder (which has hard stops which is why I use it as left/right buttons) you change the current "page" of hotcues. As it can be ease to loose track which page you are on, the buttons on the first row will briefly show which page you are currently on after you changed the page. So if I was on hotcue page one and turned the encoder one step to the left, all the light on second row would represent the new hotcue buttons and the second button of the first row would light up for 1s.
TLDR: It displays the current hotcuepage index within the upper row of the buttongrid.

File: res/controllers/Numark-N4-4channel-scripts.js [hunk 5]
```diff
@@ -0,0 +1,674 @@
...
+      this.input = function (channel, control, value, status, group) {
+        this.currentLoopSizeIndex=
+        Math.max(
+          Math.min(
+            this.currentLoopSizeIndex+(value===0x01?1:-1),
+            NumarkN4.loopSizes.length
+          ),
+          0
+        );
+        this.inSetValue(NumarkN4.loopSizes[this.currentLoopSizeIndex]);
```
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/Numark-N4-4channel-scripts.js]: Looking at the Code, it was probably my intention back when I wrote the code but your point makes more sense. I'll change that to your halve/double approach.

File: res/controllers/Numark-N4-4channel-scripts.js [hunk 6]
```diff
@@ -0,0 +1,674 @@
...
+
+  // NOTE: [Mixer Profile] is not a documented group. Taken from script.crossfaderCurve
+  // BUG: Help on Mixxx forum is needed
+  this.changeCrossfaderContour = new components.Button({
+    midi: [0x90,0x4B],
+    group: "[Mixer Profile]",
+    inKey: "xFaderMode",
+    input: function (channel, control, value, status, group) {
+      if (this.isPress(channel,control, value, status)) {
+        engine.setValue("[Mixer Profile]","xFaderCurve",2); // REVIEW: value yet to review
```
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/Numark-N4-4channel-scripts.js]: I'm not quite sure about this code as I don't want to override the users preferred cf parameters within normal mode. I'll rewrite some of that to save the users preferred parameters beforehand and restore them on shutdown/normalmode.

File: res/controllers/Numark-N4-4channel-scripts.js [hunk 7]
```diff
@@ -0,0 +1,674 @@
...
+        engine.scratchTick(channel,value);
+      } else {
+        this.inSetValue(value);
+      }
+    },
+  });
+
+  this.manageChannelIndicator = function () {
+    this.alternating=!this.alternating; //mimics a static variable
+    midi.sendShortMsg(0xB0,0x1D+channel,((engine.getValue(theDeck.group,"playposition")>NumarkN4.warnAfterPosition)&&this.alternating?0x7F:0x0));
```
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/Numark-N4-4channel-scripts.js]: Yes probably. The best case scenario would be if I had Access to ([Waveform],"EndOfTrackWarningTime) but that is a only part of the config and not a ControlObject.

File: res/controllers/Numark-N4-4channel-scripts.js [hunk 8]
```diff
@@ -0,0 +1,674 @@
...
+    flickerSafetyTimeout: true,
+    input: function (channel, control, value, status, group) {
+      if (this.flickerSafetyTimeout) {
+        this.flickerSafetyTimeout=false;
+        value/=0x7F;
+        if (this.inGetParameter()!==value){
+          this.inSetParameter(value);
+        }
+        engine.beginTimer(100,function () {
+          this.flickerSafetyTimeout=true;
```
[the remark this answers is not in the document]
Comment (Be-ing) [reply res/controllers/Numark-N4-4channel-scripts.js]: Okay, yeah please document the strange behavior of the controller in the code.

**Comment by Be-ing** [comment]
I notice the name of the XML file is "Numark N4 4channel.midi.xml", but the name of the controller is simply "Numark N4". Could you rename the XML file to overwrite the old mapping?

File: res/controllers/Numark-N4-4channel-scripts.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/Numark-N4-4channel-scripts.js]: Should I keep the beatjump code or should I implement my own double/halve logic?

**Comment by Swiftb0y** [comment]
I have incorporated all requested changes from the review but I wasn't able to test the them yet as I don't have access to the controller right now. so **DO NOT MERGE** yet. I'll be able to test them tomorrow.

File: res/controllers/Numark-N4-scripts.js [hunk 9]
```diff
@@ -649,5 +663,7 @@ NumarkN4.shutdown = function () {
     // View Definition of Array for explanation.
     NumarkN4.Decks[i].shutdown();
   }
+	// reset crossfader parameters to user preferences.
```
Comment (Be-ing) [inline res/controllers/Numark-N4-scripts.js]: use spaces instead of a tab

File: res/controllers/Numark-N4-scripts.js [hunk 10]
```diff
@@ -639,6 +652,7 @@ NumarkN4.Deck = function (channel) {
      this.playButton.send(0);
      this.shiftButton.send(0);
      midi.sendShortMsg(0xB0,0x1D+channel,0); // turn off small triangle above LOAD button.
+
```
Comment (Be-ing) [inline res/controllers/Numark-N4-scripts.js]: extra newline

**Review by Be-ing — CHANGES_REQUESTED** [review]

**Comment by Swiftb0y** [comment]
@Be-ing, The mapping is now finally ready to merge. I still have to updated the Wiki with the latest tweaks but the script files are done.

**Comment by Be-ing** [comment]
Great, thanks!
